### ISSUE TITLE
New rule: Exceptions should not be thrown in finally blocks

### ISSUE THREAD

**Issue by meziantou** [meziantou/Meziantou.Analyzer#80]

**Comment by louis-z** [comment]
I _(quickly)_ implemented a first version of the `DoNotThrowExceptionFromFinallyBlockAnalyzer`. Is it supposed to detect only explicit `throw` statements or must it also determine whether any call made from a `finally` block is likely to throw?

**Comment by meziantou** [comment]
We cannot do advanced checks. So, I think it should only detect direct throw statements. Maybe we can skip catched exceptions.

````
try
{
}
finally
{
    throw new Exception(); // report

    try
    {
         throw new Exception(); // ok
    }
    catch
    {
    }

    try
    {
         throw new Exception(); // report?
    }
    catch (SpecificException) // doesn't catch Exception
    {
    }
}
````

**Comment by louis-z** [comment]
And of course, I'll need to be smarter than just rely on `SemanticModel.AnalyzeControlFlow` and check if `EndPointIsReachable` is false... 😉

**Comment by meziantou** [comment]
To be honest, I think we you just detect any throw statement in a finally block. They all seems wrong to me.
So, an implementation based on IOperation should be fine.
